src/Microsoft.ServiceFabric.Actors/Runtime/MigrationWorker.cs

```diff
@@ -0,0 +1,232 @@
+            {
+                var dataToMigrate = await this.GetEnumeratedDataFromKvsAsync(phase, startSN, endSN - startSN);
+                await this.stateProvider.SaveStatetoRCAsync(dataToMigrate, cancellationToken);
+                await this.UpdateMetadataOnCompletion(phase, workerIdentifier, endSN);
+            }
+            else
+            {
+                var lastUpdatedSN = startSN - 1;
+                var endSNForChunk = lastUpdatedSN + this.itemsPerEnumeration;
+                while (endSNForChunk <= endSN)
+                {
+                    var dataToMigrate = await this.GetEnumeratedDataFromKvsAsync(phase, lastUpdatedSN + 1, this.itemsPerEnumeration);
+                    await this.stateProvider.SaveStatetoRCAsync(dataToMigrate, cancellationToken);
+                    lastUpdatedSN = endSNForChunk;
+                    endSNForChunk += this.itemsPerEnumeration;
+                    await this.UpdateMetadataForLastAppliedSN(phase, workerIdentifier, lastUpdatedSN);
+                }
+
+                if (lastUpdatedSN < endSN)
+                {
+                    var dataToMigrate = await this.GetEnumeratedDataFromKvsAsync(phase, lastUpdatedSN + 1, endSN - lastUpdatedSN);
+                    await this.stateProvider.SaveStatetoRCAsync(dataToMigrate, cancellationToken);
+                }
+
+                await this.UpdateMetadataOnCompletion(phase, workerIdentifier, endSN);
+            }
+        }
+
+        private async Task<List<KeyValuePair<string, byte[]>>> GetEnumeratedDataFromKvsAsync(MigrationPhase phase, long start, long enumerationSize)
+        {
+            var apiName = phase == MigrationPhase.Copy ? MigrationConstants.EnumeratebySNEndpoint : MigrationConstants.EnumerateKeysAndTombstonesEndpoint;
+            bool includeDeletes = phase == MigrationPhase.Copy;
+            var requestserializer = new DataContractSerializer(typeof(EnumerationRequest));
+            var keyvaluepairserializer = new DataContractSerializer(typeof(List<KeyValuePair<string, byte[]>>));
+
+            var enumerationRequestContent = this.CreateEnumerationRequestObject(start, enumerationSize, includeDeletes);
+
+            using var memoryStream = new MemoryStream();
+            var binaryWriter = XmlDictionaryWriter.CreateBinaryWriter(memoryStream);
+            requestserializer.WriteObject(binaryWriter, enumerationRequestContent);
+            binaryWriter.Flush();
+
+            var byteArray = memoryStream.ToArray();
+            var content = new ByteArrayContent(byteArray);
+            var kvsApiRequest = new HttpRequestMessage
+            {
+                Method = HttpMethod.Get,
+                RequestUri = new Uri(this.endpoint + apiName),
+                Content = content,
+            };
+
+            var response = await Client.SendAsync(kvsApiRequest);
+            response.EnsureSuccessStatusCode();
+            var stream = await response.Content.ReadAsStreamAsync();
```

Comment:
You should not read the entire stream at once. You should read the stream chunk by chunk.
The stream will contain array of List<KeyValuePair>
https://github.com/microsoft/service-fabric-services-and-actors-dotnet/blob/kvs-rc-migration/src/Microsoft.ServiceFabric.Actors.Migration/Extensions/KvsActorStateProviderExtensionHelper.cs